feat: add support for token helper

bruderstein opened the pull request:
Add support for token helper, a command line tool to obtain a token.

A token helper is an executable, set in the user's `.npmrc` which
outputs an auth token. This can be used in situations where the
`authToken` is not a constant value, but is something that refreshes
regularly, where a script or other tool can use an existing refresh
token to obtain a new access token.

The configuration for the path to the helper must be an absolute path,
with no arguments. In order to be secure, it is _only_ permitted to set
this value in the user `.npmrc`, otherwise a project could place a value
in a project local `.npmrc` and run arbitrary executables. For this
reason, the userConfig is added as a third parameter to the
function, to enable only allowing this value from that `.npmrc`.

A similar feature is available in many similar tools. The implementation
here is modelled after the `vault` (vaultproject.io) implementation
- https://github.com/hashicorp/vault/blob/main/command/token/helper_external.go

I have a PR ready to go for pnpm itself to use this. I'll raise that shortly, even though that will rely on this, and require the new version of this package to be published.

Commit:
feat: add support for token helper

Add support for token helper, a command line tool to obtain a token.

A token helper is an executable, set in the user's `.npmrc` which
outputs an auth token. This can be used in situations where the
`authToken` is not a constant value, but is something that refreshes
regularly, where a script or other tool can use an existing refresh
token to obtain a new access token.

The configuration for the path to the helper must be an absolute path,
with no arguments. In order to be secure, it is _only_ permitted to set
this value in the user `.npmrc`, otherwise a project could place a value
in a project local `.npmrc` and run arbitrary executables. For this
reason, the userConfig is added as a third parameter to the
function, to enable only allowing this value from that `.npmrc`.

A similar feature is available in many similar tools. The implementation
here is modelled after the `vault` (vaultproject.io) implementation
- https://github.com/hashicorp/vault/blob/main/command/token/helper_external.go

zkochan commented on index.js:
```diff
@@ -26,6 +29,21 @@ function getScopedCredentials (nerfed, scope, config) {
+  if (userConfig && userConfig[`${scope}tokenHelper`]) {
+    const helper = userConfig[`${scope}tokenHelper`]
+    if (!path.isAbsolute(helper) || !fs.existsSync(helper)) {
+      throw new Error(`${scope}tokenHelper must be an absolute path, without arguments`)
```
Install `@pnpm/error` as a dependency and throw an error with an error code. Something like
```suggestion
      throw new PnpmError('BAD_TOKEN_HELPER_PATH', `${scope}tokenHelper must be an absolute path, without arguments`)
```

Same for the below error.

bruderstein replied:
Done

zkochan commented on test-exec.js:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env node
```
this won't work on Windows.

bruderstein replied:
I'll get this fixed. Thanks for the super fast review.

bruderstein replied:
Done.

bruderstein committed:
test: fix tokenHelper tests on windows

The windows tests now run a batch file which outputs the token
or exits with an exit code.

Commit:
fix: use @pnpm/error for throwing errors

bruderstein commented:
@zkochan applied the changes from the comments, tested on windows, linux + mac.

I tried to make the OS dependent parts such that coverage would not be affected, so it's still at 100% coverage.
